bundles/org.openhab.binding.shelly/src/main/java/org/openhab/binding/shelly/internal/api/ShellyHttpApi.java

```diff
@@ -379,154 +348,216 @@ private void setRelayEvents() throws IOException {
+                    // Sensor in sleep mode or
+                    // API exception for non-battery device
+                    throw e; // non-timeout exception
+                }
+
+                timeoutErrors++; // count the retries
+                logger.debug("{}: {}, retry #{}", thingName, e.toString(), timeoutErrors);
+                try {
+                    apiResult = innerRequest(HttpMethod.GET, uri);
+                    timeoutsRecovered++;
+                    logger.debug("{}: API timeout #{}/{} recovered ({})", thingName, timeoutErrors, timeoutsRecovered,
+                            apiResult.getUrl());
+                } catch (ShellyApiException e2) {
+                    if (retries <= 1) {
+                        throw new ShellyApiException(e2);
+                    }
                 }
             }
         }
-        return result;
+        throw new ShellyApiException("Inconsistent API result or Timeout"); // successful
     }
 
-    private String innerRequest(String uri) throws IOException {
-        String httpResponse = "ERROR";
+    private ShellyApiResult innerRequest(HttpMethod method, String uri) throws ShellyApiException {
+        Request request = null;
         String url = "http://" + config.deviceIp + uri;
-        logger.trace("{}: HTTP GET for {}", thingName, url);
+        ShellyApiResult apiResult = new ShellyApiResult(method.toString(), url);
 
-        Properties headers = new Properties();
-        if (!config.userId.isEmpty()) {
-            String value = config.userId + ":" + config.password;
-            headers.put(HTTP_HEADER_AUTH,
-                    HTTP_AUTH_TYPE_BASIC + " " + Base64.getEncoder().encodeToString(value.getBytes()));
-        }
+        try {
+            request = httpClient.newRequest(url).method(method.toString()).timeout(SHELLY_API_TIMEOUT_MS,
+                    TimeUnit.MILLISECONDS);
 
-        httpResponse = HttpUtil.executeUrl(HttpMethod.GET, url, headers, null, "", SHELLY_API_TIMEOUT_MS);
-        Validate.notNull(httpResponse, "httpResponse must not be null");
-        // all api responses are returning the result in Json format. If we are getting
-        // something else it must
-        // be an error message, e.g. http result code
-        if (httpResponse.contains(APIERR_HTTP_401_UNAUTHORIZED)) {
-            throw new IOException(
-                    APIERR_HTTP_401_UNAUTHORIZED + ", set/correct userid and password in the thing/binding config");
-        }
-        if (!httpResponse.startsWith("{") && !httpResponse.startsWith("[")) {
-            throw new IOException("Unexpected http response: " + httpResponse);
+            if (!config.userId.isEmpty()) {
+                String value = config.userId + ":" + config.password;
+                request.header(HTTP_HEADER_AUTH,
+                        HTTP_AUTH_TYPE_BASIC + " " + Base64.getEncoder().encodeToString(value.getBytes()));
```

Comment:
You should specify a charset.

Reply:
changed, now using getBytes(StandardCharsets.UTF_8) and
public static final String CONTENT_TYPE_JSON = "application/json; charset=UTF-8";
ok?